File: src/platform/posix/mem.c
/*
 * Copyright 2014-2016, Björn Ståhl
 * License: 3-Clause BSD, see COPYING file in arcan source repository.
 * Reference: http://arcan-fe.com
 */

/*
 * This is mostly a stub implementation still, serious attempts
 * at getting a strong and efficient implementation is slated for
 * 0.8-0.9.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include <stdbool.h>
#include <signal.h>
#include <unistd.h>

#include <sys/mman.h>

#include "arcan_math.h"
#include "arcan_general.h"
#include "arcan_video.h"

#ifdef MADV_DONTDUMP
#define NO_DUMPFLAG MADV_DONTDUMP
#else
#ifndef MADV_NOCORE
#define NO_DUMPFLAG 0
#else
#define	NO_DUMPFLAG MADV_NOCORE
#endif
#endif

#ifndef REALLOC_STEP
#define REALLOC_STEP 16
#endif

struct mempool_meta {
/*	mempool_hook_t alloc;
	  mempool_hook_t free; */

	size_t alloc_cnt;
	size_t dealloc_cnt;
	size_t in_use;
	size_t monitor_sz;
	size_t n_pages;
};

/* pool behaviors:
 * [ SENSITIVE is always a special case ]
 *   |-> pages will remain mapped in dumps, but data will be
 *       written over with the guard byte.
 *
 * ARCAN_MEM_VBUFFER =>
 *  - assume large continous blocks are beneficial,
 *    initialize with 0xff so that rendering an uninitialized
 *    block will yield something visible, tightly pack within
 *    pools, don't use any guard or additional tracking.
 *
 *    hope is that MEM_VBUFFER + READONLY can be used with
 *    future low-level graphics interfaces to directly map
 *    and manage GPU resources.
 *
 *    MEM_VBUFFER + BZERO uses the _video.h RGBA packing and
 *    PIXEL_BPP to set alpha channel to fully on.
 *
 * ARCAN_MEM_VSTRUCT =>
 *  - scratch-page + tightly packed ring-buffer, 16-byte align
 *    base, guardint and checksum as the buffer is tracked.
 *
 * ARCAN_MEM_EXTSTRUCT =>
 *  - same as VSTRUCT but different pool
 *
 * ARCAN_MEM_ABUFFER =>
 *  - similar to VBUFFER but smaller blocks (multiples of 64k)
 *    and no 0xff initialization
 *
 * ARCAN_MEM_STRINGBUF =>
 *  - plan is to require additional meta (i.e. encoding)
 *    aggressively verify checksums (as we know there's little
 *    data but high risk), dynamically tagged readonly/ sensitive,
 *    also gets all access tracked (in more paranoid settings,
 *    madv. use should also be covered but we don't have the
 *    mechanisms in place for that) by getting the pagefaults
 *    and checking source.
 *
 * ARCAN_MEM_VTAG/ATAG =>
 *  - similar to EXTSTRUCT we know that the source is not
 *    connected to external behaviors
 *
 * ARCAN_MEM_BINDING =>
 *  - 64k pages managed by additional memory managers,
 *    e.g. allocation calls to virtual machines. Possibly
 *    connected to a JITing source(!),
 *
 * ARCAN_MEM_MODELDATA =>
 *  - similar to VBUFFER but smaller allocation blocks
 *
 * ARCAN_MEM_THREADCTX =>
 *   - few allocations (should correlate to number of threads)
 *   - assumed shorter life-span
 *   - guard pages separate each allocation
 */

int system_page_size = 4096;

/*
 * map initial pools, pre-fill some video buffers,
 * get limits and assert that our build-time minimal
 * and maximal settings fit.
 */
void arcan_mem_init()
{

}

/*
 * there should essentially be NO memory blocks marked
 * TEMPORARY or SENSITIVE (NON VIDEO/AUDIO) alive at this
 * point, use the tick point to check and trap as leaks.
 */
void arcan_mem_tick()
{
}

/*static void sigsegv_hand(int sig, siginfo_t* si, void* unused)
{
*/
/*
 * sweep SENSITIVE marked areas and replace with the guard pattern
 *
 */

/*
 * unmap VBUFFER, ABUFFER, MODELDATA
 */

/*
 * Propagate the segmentation fault so the core gets dumped.
 *
 * Should do an integrity sweep and exfiltrate additional
 * crash information
	struct sigaction sa;
 */

void* arcan_alloc_mem(size_t nb,
	enum arcan_memtypes type, enum arcan_memhint hint, enum arcan_memalign align)
{
	void* rptr = NULL;
	size_t header_sz = 0;
	size_t footer_sz = 0;
	size_t padding_sz = 0;
	size_t total;

	switch (type){
/* this is not safe to _free at the moment since we don't track metadata yet,
 * the refactor adding obsd/musl-new heap allocator should take this into account */
	case ARCAN_MEM_SHARED:
		rptr = mmap(NULL, system_page_size,
			PROT_READ | PROT_WRITE,
			MAP_SHARED | MAP_ANONYMOUS, -1, 0
		);
		total = system_page_size;

	break;
	case ARCAN_MEM_BINDING:
	case ARCAN_MEM_THREADCTX:
	case ARCAN_MEM_VBUFFER:
	case ARCAN_MEM_ABUFFER:
	case ARCAN_MEM_MODELDATA:
	case ARCAN_MEM_VSTRUCT:
	case ARCAN_MEM_EXTSTRUCT:
	case ARCAN_MEM_STRINGBUF:
	case ARCAN_MEM_VTAG:
	case ARCAN_MEM_ATAG:

		total = header_sz + footer_sz + padding_sz + nb;

		switch(align){
		case ARCAN_MEMALIGN_NATURAL:
			rptr = malloc(total);
		break;

		case ARCAN_MEMALIGN_PAGE:
			if (-1 == posix_memalign(&rptr, system_page_size, total))
				rptr = NULL;
		break;

		case ARCAN_MEMALIGN_SIMD:
			if (-1 == posix_memalign(&rptr, 16, total))
				rptr = NULL;
		break;
		}
	break;

	case ARCAN_MEM_ENDMARKER:
	default:
			abort();
	}

	if (!rptr){
		if ((hint & ARCAN_MEM_NONFATAL) == 0)
			arcan_fatal("arcan_alloc_mem(), out of memory.\n");
		return NULL;
	}

/*
 * Post-alloc hooks
 */
	int madvflag = 0;

	switch(type){
	case ARCAN_MEM_VBUFFER:
	case ARCAN_MEM_ABUFFER:
	case ARCAN_MEM_MODELDATA:
		madvflag |= NO_DUMPFLAG;
	break;

	default:
	break;
	}

	if (hint & ARCAN_MEM_SENSITIVE)
		madvflag |= NO_DUMPFLAG;

	if (madvflag)
		madvise(rptr, total, madvflag);

	if (hint & ARCAN_MEM_BZERO){
		if (type == ARCAN_MEM_VBUFFER){
			av_pixel* buf = (av_pixel*) rptr;
			for (size_t i = 0; i < nb; i += sizeof(av_pixel))
				*buf++ = RGBA(0, 0, 0, 255);
		}
		else
			memset(rptr, '\0', total);
	}

	return rptr;
}

void arcan_mem_growarr(struct arcan_strarr* res)
{
/* _alloc functions lacks a grow at the moment,
 * after that this should ofc. be replaced. */
	char** newbuf = arcan_alloc_mem(
		(res->limit + REALLOC_STEP) * sizeof(char*),
		ARCAN_MEM_STRINGBUF, ARCAN_MEM_BZERO, ARCAN_MEMALIGN_NATURAL
	);

	if (res->data)
		memcpy(newbuf, res->data, res->limit * sizeof(char*));
	arcan_mem_free(res->data);
	res->data= newbuf;
	res->limit += REALLOC_STEP;
}

void arcan_mem_freearr(struct arcan_strarr* res)
{
	if (!res || !res->data)
		return;

	char** cptr = res->data;
	while (cptr && *cptr)
		arcan_mem_free(*cptr++);

	arcan_mem_free(res->data);

	memset(res, '\0', sizeof(struct arcan_strarr));
}

/*
 * Allocate memory intended for read-only or exec use (JIT, ...)
 */
void* arcan_alloc_fillmem(const void* data,
	size_t ds,
	enum arcan_memtypes type,
	enum arcan_memhint hint,
	enum arcan_memalign align)
{
	void* buf = arcan_alloc_mem(ds, type, hint, align);

	if (!buf)
		return NULL;

	memcpy(buf, data, ds);

	return buf;
}

void arcan_mem_free(void* inptr)
{
/* lock then free */
/* depending on type and flag, verify integrity,
 * then cleanup. VBUFFER for instance doesn't
 * automatically shrink, but rather reset and flag
 * as unused */

	free(inptr);
}

File: src/platform/posix/mem.zig
// Zig port of posix/mem.c
// Type/use-hinted memory (de-)allocation routines.
// Currently a stub implementation — serious hardening slated for 0.8-0.9.

const builtin = @import("builtin");
const is_freestanding = (builtin.os.tag == .freestanding);

const c = @import("shmif_types");

// MADV_DONTDUMP / MADV_NOCORE for excluding pages from core dumps.
// Linux defines MADV_DONTDUMP, BSDs use MADV_NOCORE.
const NO_DUMPFLAG: c_int = if (is_freestanding) 0 else c.MADV_DONTDUMP;

const REALLOC_STEP: usize = 16;

extern fn arcan_fatal(fmt: [*c]const u8, ...) callconv(.c) void;

// MAY-258: freestanding (EL2 bare-metal) engine heap, backed by the seL4
// BootInfo-sourced free-list allocator in src/sel4-zig/sel4_heap.zig. Reached by
// C-ABI extern symbol (these live in the root module of the boot payload, this
// file is compiled into the engine module). On hosted/posix builds these
// symbols are never referenced (the is_freestanding branches are dead).
extern fn el2_heap_alloc(size: usize, alignment: usize) ?*anyopaque;
extern fn el2_heap_free(ptr: ?*anyopaque) void;

// m1n1 uses a 16K page (Apple Silicon). NATURAL/SIMD map to the heap's 16-byte
// minimum; PAGE maps to the 16K page so PAGE-aligned engine allocs are honored.
const FS_ALIGN_NATURAL: usize = 16;
const FS_ALIGN_SIMD: usize = 16;
const FS_ALIGN_PAGE: usize = 16384;

export var system_page_size: c_int = 4096;

/// Initialize memory pools. Currently a stub.
export fn arcan_mem_init() void {}

/// Tick-based memory housekeeping. Currently a stub.
export fn arcan_mem_tick() void {}

const enum_arcan_memtypes = c.enum_arcan_memtypes;
const enum_arcan_memhint = c.enum_arcan_memhint;
const enum_arcan_memalign = c.enum_arcan_memalign;

export fn arcan_alloc_mem(
    nb: usize,
    mem_type: enum_arcan_memtypes,
    hint: enum_arcan_memhint,
    alignment: enum_arcan_memalign,
) ?*anyopaque {
    if (is_freestanding) {
        // EL2 bare-metal: route every type through the seL4-sourced engine heap.
        // ENDMARKER / unknown types still fatal (matches the posix path's else).
        switch (mem_type) {
            c.ARCAN_MEM_SHARED,
            c.ARCAN_MEM_BINDING,
            c.ARCAN_MEM_THREADCTX,
            c.ARCAN_MEM_VBUFFER,
            c.ARCAN_MEM_ABUFFER,
            c.ARCAN_MEM_MODELDATA,
            c.ARCAN_MEM_VSTRUCT,
            c.ARCAN_MEM_EXTSTRUCT,
            c.ARCAN_MEM_STRINGBUF,
            c.ARCAN_MEM_VTAG,
            c.ARCAN_MEM_ATAG,
            => {},
            else => arcan_fatal("arcan_alloc_mem(): invalid memtype (freestanding)\n"),
        }

        const al: usize = switch (alignment) {
            c.ARCAN_MEMALIGN_PAGE => FS_ALIGN_PAGE,
            c.ARCAN_MEMALIGN_SIMD => FS_ALIGN_SIMD,
            else => FS_ALIGN_NATURAL, // NATURAL + any unknown
        };

        // el2_heap_alloc zeroes the block, satisfying ARCAN_MEM_BZERO for every
        // type. (The posix path fills VBUFFER with opaque-black RGBA(0,0,0,255)
        // rather than raw zero; we keep raw-zero here — the engine's first allocs
        // on the bringup path are VSTRUCT, not VBUFFER, and zeroed VBUFFER is a
        // transparent-black rather than opaque-black canvas, a display nicety not
        // needed for correctness. RGBA() isn't exposed by shmif_types anyway.)
        const rptr = el2_heap_alloc(nb, al);
        if (rptr == null) {
            // Honor ARCAN_MEM_NONFATAL: callers like populate_vstore handle null.
            if ((@as(c_uint, @bitCast(hint)) & @as(c_uint, @bitCast(c.ARCAN_MEM_NONFATAL))) == 0) {
                arcan_fatal("arcan_alloc_mem(), out of memory (freestanding heap).\n");
            }
            return null;
        }
        return rptr;
    }
    var rptr: ?*anyopaque = null;
    const header_sz: usize = 0;
    const footer_sz: usize = 0;
    const padding_sz: usize = 0;
    var total: usize = 0;

    switch (mem_type) {
        // SHARED: mmap an anonymous shared page (not safe to _free yet)
        c.ARCAN_MEM_SHARED => {
            const page_sz: usize = @intCast(system_page_size);
            const ptr = c.mmap(
                null,
                page_sz,
                c.PROT_READ | c.PROT_WRITE,
                c.MAP_SHARED | c.MAP_ANONYMOUS,
                -1,
                0,
            );
            if (ptr != c.MAP_FAILED) {
                rptr = ptr;
            }
            total = page_sz;
        },

        // All other valid types: heap allocation with alignment
        c.ARCAN_MEM_BINDING,
        c.ARCAN_MEM_THREADCTX,
        c.ARCAN_MEM_VBUFFER,
        c.ARCAN_MEM_ABUFFER,
        c.ARCAN_MEM_MODELDATA,
        c.ARCAN_MEM_VSTRUCT,
        c.ARCAN_MEM_EXTSTRUCT,
        c.ARCAN_MEM_STRINGBUF,
        c.ARCAN_MEM_VTAG,
        c.ARCAN_MEM_ATAG,
        => {
            total = header_sz + footer_sz + padding_sz + nb;

            switch (alignment) {
                c.ARCAN_MEMALIGN_NATURAL => {
                    rptr = c.malloc(total);
                },
                c.ARCAN_MEMALIGN_PAGE => {
                    var aligned: ?*anyopaque = null;
                    if (c.posix_memalign(&aligned, @intCast(system_page_size), total) == 0) {
                        rptr = aligned;
                    }
                },
                c.ARCAN_MEMALIGN_SIMD => {
                    var aligned: ?*anyopaque = null;
                    if (c.posix_memalign(&aligned, 16, total) == 0) {
                        rptr = aligned;
                    }
                },
                else => {},
            }
        },

        // ENDMARKER or unknown: fatal
        else => {
            c.abort();
        },
    }

    if (rptr == null) {
        if ((@as(c_uint, @bitCast(hint)) & @as(c_uint, @bitCast(c.ARCAN_MEM_NONFATAL))) == 0) {
            arcan_fatal("arcan_alloc_mem(), out of memory.\n");
        }
        return null;
    }

    // Post-alloc hooks: madvise for buffer types and sensitive data
    var madvflag: c_int = 0;

    switch (mem_type) {
        c.ARCAN_MEM_VBUFFER,
        c.ARCAN_MEM_ABUFFER,
        c.ARCAN_MEM_MODELDATA,
        => {
            madvflag |= NO_DUMPFLAG;
        },
        else => {},
    }

    if ((@as(c_uint, @bitCast(hint)) & @as(c_uint, @bitCast(c.ARCAN_MEM_SENSITIVE))) != 0) {
        madvflag |= NO_DUMPFLAG;
    }

    if (madvflag != 0) {
        _ = c.madvise(rptr, total, madvflag);
    }

    if ((@as(c_uint, @bitCast(hint)) & @as(c_uint, @bitCast(c.ARCAN_MEM_BZERO))) != 0) {
        if (mem_type == c.ARCAN_MEM_VBUFFER) {
            // VBUFFER zero: fill with RGBA(0,0,0,255) — opaque black pixels
            const buf_ptr: [*]c.av_pixel = @ptrCast(@alignCast(rptr));
            const pixel_count = nb / @sizeOf(c.av_pixel);
            const opaque_black = c.RGBA(0, 0, 0, 255);
            for (0..pixel_count) |i| {
                buf_ptr[i] = opaque_black;
            }
        } else {
            _ = c.memset(rptr, 0, total);
        }
    }

    return rptr;
}

const struct_arcan_strarr = if (is_freestanding) anyopaque else c.struct_arcan_strarr;

export fn arcan_mem_growarr(res: *struct_arcan_strarr) void {
    if (is_freestanding) return;
    const new_limit = res.limit + REALLOC_STEP;
    const newbuf: [*c][*c]u8 = @ptrCast(@alignCast(arcan_alloc_mem(
        new_limit * @sizeOf([*c]u8),
        c.ARCAN_MEM_STRINGBUF,
        c.ARCAN_MEM_BZERO,
        c.ARCAN_MEMALIGN_NATURAL,
    )));

    if (res.unnamed_0.data != null) {
        _ = c.memcpy(
            @as(?*anyopaque, @ptrCast(newbuf)),
            @as(?*const anyopaque, @ptrCast(res.unnamed_0.data)),
            res.limit * @sizeOf([*c]u8),
        );
    }
    arcan_mem_free(@as(?*anyopaque, @ptrCast(res.unnamed_0.data)));
    res.unnamed_0.data = newbuf;
    res.limit = new_limit;
}

export fn arcan_mem_freearr(res: ?*struct_arcan_strarr) void {
    if (is_freestanding) return;
    const r = res orelse return;
    if (r.unnamed_0.data == null) return;

    var cptr: [*c][*c]u8 = r.unnamed_0.data;
    while (cptr[0] != null) {
        arcan_mem_free(@as(?*anyopaque, @ptrCast(cptr[0])));
        cptr += 1;
    }

    arcan_mem_free(@as(?*anyopaque, @ptrCast(r.unnamed_0.data)));

    _ = c.memset(
        @as(?*anyopaque, @ptrCast(r)),
        0,
        @sizeOf(c.struct_arcan_strarr),
    );
}

/// Allocate and fill memory from a source buffer.
export fn arcan_alloc_fillmem(
    data: ?*const anyopaque,
    ds: usize,
    mem_type: enum_arcan_memtypes,
    hint: enum_arcan_memhint,
    alignment: enum_arcan_memalign,
) ?*anyopaque {
    if (is_freestanding) {
        const buf = arcan_alloc_mem(ds, mem_type, hint, alignment) orelse return null;
        if (data) |src| {
            const dst: [*]u8 = @ptrCast(buf);
            const src_b: [*]const u8 = @ptrCast(src);
            @memcpy(dst[0..ds], src_b[0..ds]);
        }
        return buf;
    }
    const buf = arcan_alloc_mem(ds, mem_type, hint, alignment) orelse return null;
    _ = c.memcpy(buf, data, ds);
    return buf;
}

export fn arcan_mem_free(inptr: ?*anyopaque) void {
    if (is_freestanding) {
        // free + coalesce in the seL4-sourced heap; no-op on null / out-of-arena.
        el2_heap_free(inptr);
        return;
    }
    c.free(inptr);
}
